src/vm/analysis/arithmetic_checker/mod.rs

```diff
@@ -0,0 +1,276 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use vm::analysis::types::{AnalysisPass, ContractAnalysis};
+use vm::functions::define::{DefineFunctions, DefineFunctionsParsed};
+use vm::functions::tuples;
+use vm::functions::NativeFunctions;
+use vm::representations::SymbolicExpressionType::{
+    Atom, AtomValue, Field, List, LiteralValue, TraitReference,
+};
+use vm::representations::{ClarityName, SymbolicExpression, SymbolicExpressionType};
+use vm::types::{parse_name_type_pairs, PrincipalData, TupleTypeSignature, TypeSignature, Value};
+
+use std::collections::HashMap;
+use vm::variables::NativeVariables;
+
+pub use super::errors::{
+    check_argument_count, check_arguments_at_least, CheckError, CheckErrors, CheckResult,
+};
+use super::AnalysisDatabase;
+
+#[cfg(test)]
+mod tests;
+
+///
+/// A static-analysis pass that checks whether or not
+///  a proposed cost-function defining contract is allowable.
+/// Cost-function defining contracts may not use contract-call,
+///  any database operations, or iterating operations (e.g., list
+///  operations)
+///
+pub struct ArithmeticOnlyChecker();
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum Error {
+    DefineTypeForbidden(DefineFunctions),
+    VariableForbidden(NativeVariables),
+    FunctionNotPermitted(NativeFunctions),
+    TraitReferencesForbidden,
+    UnexpectedContractStructure,
+}
+
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        None
+    }
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl ArithmeticOnlyChecker {
+    pub fn check_contract_cost_eligible(contract_analysis: &mut ContractAnalysis) {
+        let is_eligible = ArithmeticOnlyChecker::run(contract_analysis).is_ok();
+        contract_analysis.is_cost_contract_eligible = is_eligible;
+    }
+
+    pub fn run(contract_analysis: &ContractAnalysis) -> Result<(), Error> {
+        let checker = ArithmeticOnlyChecker();
+        for exp in contract_analysis.expressions.iter() {
+            checker.check_top_levels(&exp)?;
+        }
+
+        Ok(())
+    }
+
+    fn check_define_function(
+        &self,
+        _signature: &[SymbolicExpression],
+        body: &SymbolicExpression,
+    ) -> Result<(), Error> {
+        self.check_expression(body)
+    }
+
+    fn check_top_levels(&self, expr: &SymbolicExpression) -> Result<(), Error> {
+        use vm::functions::define::DefineFunctionsParsed::*;
+        if let Some(define_type) = DefineFunctionsParsed::try_parse(expr)
+            .map_err(|_| Error::UnexpectedContractStructure)?
+        {
+            match define_type {
+                // The _arguments_ to Constant, PersistedVariable, FT defines must be checked to ensure that
```

Comment:
Comment copypasta?  It doesn't look like you're doing anything with `PersistedVariable` and `FT`.

Reply:
Ah, I think this comment may have been left-over from a first implementation of this pass to make it more generic, will clean up the comments.

Reply:
Cleaned up in https://github.com/blockstack/stacks-blockchain/pull/2143/commits/c03277838d1e045d9f6c72c2e13e861b90819369